File: Services/pkg/domain/role.go
package domain

import (
	"context"

	"github.com/ChangerzaryX1602/SkillSync/pkg/models"

	helpers "github.com/zercle/gofiber-helpers"
)

type RoleRepository interface {
	Migrate() error
	CreateRole(ctx context.Context, role models.Role) *helpers.ResponseError
	GetRole(ctx context.Context, id uint) (*models.Role, *helpers.ResponseError)
	GetRoles(ctx context.Context, pagination models.Pagination, search models.Search) ([]models.Role, *models.Pagination, *models.Search, *helpers.ResponseError)
	GetRoleByName(ctx context.Context, name string) (*models.Role, *helpers.ResponseError)
	UpdateRole(ctx context.Context, id uint, role models.Role) *helpers.ResponseError
	DeleteRole(ctx context.Context, id uint) *helpers.ResponseError
}

type RoleService interface {
	CreateRole(ctx context.Context, role models.Role) []helpers.ResponseError
	GetRole(ctx context.Context, id uint) (*models.Role, []helpers.ResponseError)
	GetRoles(ctx context.Context, pagination models.Pagination, search models.Search) ([]models.Role, *models.Pagination, *models.Search, []helpers.ResponseError)
	UpdateRole(ctx context.Context, id uint, role models.Role) []helpers.ResponseError
	DeleteRole(ctx context.Context, id uint) []helpers.ResponseError
}

File: hono-clean-architecture/pkg/domain/role.ts
import { Role, Pagination, Search } from "../models";
import { ResponseError } from "../../internal/infrastructure/custom_error";

export interface RoleRepository {
  migrate(): Promise<void>;
  createRole(role: { name: string }): Promise<{ role: Role | null; error: ResponseError | null }>;
  getRole(id: number): Promise<{ role: Role | null; error: ResponseError | null }>;
  getRoles(
    pagination: Pagination,
    search: Search
  ): Promise<{
    roles: Role[];
    pagination: Pagination;
    search: Search;
    error: ResponseError | null;
  }>;
  updateRole(id: number, role: Partial<Role>): Promise<ResponseError | null>;
  deleteRole(id: number): Promise<ResponseError | null>;
  getRoleByName(name: string): Promise<{ role: Role | null; error: ResponseError | null }>;
}

export interface RoleService {
  createRole(role: { name: string }): Promise<ResponseError[]>;
  getRole(id: number): Promise<{ role: Role | null; errors: ResponseError[] }>;
  getRoles(
    pagination: Pagination,
    search: Search
  ): Promise<{
    roles: Role[];
    pagination: Pagination;
    search: Search;
    errors: ResponseError[];
  }>;
  updateRole(id: number, role: Partial<Role>): Promise<ResponseError[]>;
  deleteRole(id: number): Promise<ResponseError[]>;
  getRoleByName(name: string): Promise<{ role: Role | null; errors: ResponseError[] }>;
}
